### ISSUE TITLE
Feature request: min/max values for logged metrics

### ISSUE THREAD

**Issue by depthwise** [wandb/wandb#736]
Thank you for developing such a useful service. 

As a practitioner I care disproportionately about the _peak_ metrics in any given run. I.e. max mAP50 for object detection (and min classification/localization losses) or peak mIoU for segmentation. I rarely, if ever, care what the metrics are at the last step of the training run, per se. Things begin to overfit by then anyway.

W&B doesn't seem to offer a way to derive such min/max in the experiments table, or indeed anywhere its UI as far as I can tell. It'd be great (and probably not too complicated) to add this feature.

Thanks!

**Comment by vanpelt** [comment]
Hey @depthwise, thanks for the feature request.  One thing you can do immediately is log a separate metric that only updates when your metric increases or decreases, i.e. "min_loss".  What framework are you currently using?

**Comment by ayushjaiswal** [comment]
Yes, this feature would be very helpful! The parallel coordinate and parameter importance plots do not make much sense if "best" values are not used to generate them. So, it would be nice to have or be able to generate simple derived metrics like min, mean, max, etc.

**Comment by ayushjaiswal** [comment]
@vanpelt is there a plan to add this to the UI? Like I previously said, this would be extremely helpful and make parallel coordinate and importance plots more meaningful

**Comment by vanpelt** [comment]
Hey @ayushjaiswal we have plans for derived metrics but I want to make sure we understand your use case as it relates to parallel coordinates or importance.  Could you link us to an example workspace and describe how you would want it to work specifically?  

Again, in the meantime you can update the summary of a run to achieve this today:  https://docs.wandb.com/library/api/examples#update-metrics-for-a-run-after-run-finished

**Comment by ayushjaiswal** [comment]
@vanpelt is there any update on this?

**Comment by msalvaris** [comment]
@vanpelt Bump as well

**Comment by vanpelt** [comment]
Hey Guys, sorry for not responding to this sooner.   The simplest way to address this today is to log a metric that only gets updated if it improves, i.e.:

```
best_acc = 0
for e in range(epochs):
    acc = eval(...)
    best_acc = max(acc, best_acc)
    wandb.log({"best_acc": best_acc, "acc": acc})
```

We still have plans to add computed metrics in a future release, this would enable you to define new metrics using expressions like this after the fact.

**Comment by MadcowD** [comment]
Just wanted to bump this

**Comment by potipot** [comment]
Still a needed feature

**Comment by ariG23498** [comment]
I am a little confused with the thread
The initial ticket was raised for a lack of feature that could help us with finding the `min` or `max` of an experiment from the UI.
We indeed have a solution of sorting the table in an ascending or descending order of any chosen metric from the UI.
![image](https://user-images.githubusercontent.com/36856589/108347302-5e284600-7206-11eb-914b-3d0ef75864f5.png)


In the later parts of the thread, I am seeing feature requests for conditional logging where we would log only if the metrics increase or decrease.

Do you guys feel that there should be two individual issues for this? And this thread could be closed in relation to the initial ticket raised?

**Comment by depthwise** [comment]
Metrics don't always improve, and columns show the _last_ value, not the _best_ one (AKA the only one the researcher actually cares about).

**Comment by ariG23498** [comment]
Got it thanks! 😄

**Comment by kahnchana** [comment]
Hey any idea if this has this been added?

**Comment by mckinziebrandon** [comment]
Just want to point out that, although the logging solution proposed by @vanpelt works for _future_ runs, it doesn't address  what we should do for runs that have already finished. There are many of us with a _lot_ of runs that we'd like to apply this on retroactively.

**Comment by inigoval** [comment]
I also think this feature would be very useful.
It defeats the point of wandb if I have to rerun everything to get a maximum/minimum value when the loss curve has already been recorded.

**Comment by mmwebster** [comment]
Still not available? A derived metrics features would be extremely useful. Or at least being able to apply basic min/max/avg to a run's logged metric values, rather than just min/max/avg across runs' summary values

**Comment by vanpelt** [comment]
@mmwebster totally agree.  We're hard at work on a new system that will eventually support this and there are some exciting possibilities it will unlock.   Unfortunately we don't currently have a timeline for exposing this in the experiments table but we'll update this ticket when we do.

**Comment by vanpelt** [comment]
I'll also add that we've introduced a new metric api that makes this easy for future runs: https://docs.wandb.ai/ref/python/run#define_metric

For instance you could call `run.define_metric("acc", summary="max")` and this will automatically store both the latest and the max value in the summary for filtering.

**Comment by mmwebster** [comment]
@vanpelt Fantastic, looking forward to it! And thanks for mentioning `define_metric()`, that's actually really useful

**Comment by LTeder** [comment]
I've encountered this recently with my own runs. To address it, I've put together a script that updates target metrics of previous runs with best values as reported by api.run.scan_history(). [Here](https://gist.github.com/LTeder/633d400b4140e22612e997ca05e41dfc) it is.

~~Use the first revision if you don't like tqdm loading bars.~~

**Comment by miccio-dk** [comment]
I feel like I should have read this issue before starting a long sweep 😅 can't wait to see this feature implemented!

**Comment by ndrwnaguib** [comment]
> I've encountered this recently with my own runs. To address it, I've put together a script that updates target metrics of previous runs with best values as reported by api.run.scan_history(). [Here](https://gist.github.com/LTeder/633d400b4140e22612e997ca05e41dfc) it is.
> 
> ~Use the first revision if you don't like tqdm loading bars.~

Thanks for the script @LTeder, it came in really handy! BTW, I _think_ you have the [conditions reversed](https://gist.github.com/LTeder/633d400b4140e22612e997ca05e41dfc#file-wandb_update_bests-py-L30-L31) in 

```python
            if (  (func == max and bests[i] < previous[i]) or
                  (func == min and bests[i] > previous[i])):
```

Shouldn't it be:

```Python
            if (  (func == max and bests[i] > previous[i]) or
                  (func == min and bests[i] < previous[i])):
```

**Comment by LTeder** [comment]
> Thanks for the script @LTeder, it came in really handy! BTW, I _think_ you have the [conditions reversed](https://gist.github.com/LTeder/633d400b4140e22612e997ca05e41dfc#file-wandb_update_bests-py-L30-L31)

Glad you found it useful @ndrwnaguib ! It looks backwards there because the point is to print the "bad" changes during the test run. It counts and displays the value changes that wouldn't be applied (given the if-statement that follows it), but are still important to recognize. In my case, they all appeared to be rounding errors (which was about 2% of my runs, iirc)

**Comment by SteffenCzolbe** [comment]
I think this feature is finally available, with weave [expressions](https://docs.wandb.ai/ref/app/features/panels/weave#weave-expression)!

I chained an expression together like this:
`runs.map((row)` => row.history["val/loss"].min)

This creates a barplot with the lowest validation loss for each run.

**Comment by tom-andersson** [comment]
@vanpelt's `define_metric` solution works perfectly for new runs - thank you! However, I agree it would be extremely useful to be able to obtain the `min` and `max` values of the logged metrics within the UI for past runs.

**Comment by levan92** [comment]
> @vanpelt's `define_metric` solution works perfectly for new runs - thank you! However, I agree it would be extremely useful to be able to obtain the `min` and `max` values of the logged metrics within the UI for past runs.

Agreed, whether is it via weave expressions or `define_metric`, it is not directly obvious to new users. It would probably be for the best if one can easily view the min max values from the UI with no additional code added before running.

**Comment by smorad** [comment]
Just wanted to bump this once more, this would be ideal to have in the web interface. The weave solution works for plots but cannot be used when looking at tables. Furthermore, many libraries that wrap Wandb in a generic logger class will usually not provide access to the Wandb instance, preventing users from calling `define_metric`.

**Comment by janosh** [comment]
A bit tedious, I admit, but in case someone finds it helpful, [this notebook](https://github.com/janosh/tensorboard-reducer/blob/fd6c8456738/examples/wandb_integration.ipynb) of mine (scroll straight to the bottom) shows how to download WandB runs, aggregate them into `min`, `max`, `mean` (or whatever operations you want), and then re-upload the result as new runs.

![Screen Shot 2022-09-10 at 15 10 21](https://user-images.githubusercontent.com/30958850/189503575-2f5ec1c8-62f8-4285-84c2-d5e37a1b5c3e.png)

**Comment by MarkZaretskiy** [comment]
You can use the following workaround script to calculate min, max, std and whatever you want metrics if you do not want to go deep into weave expressions, etc

    import wandb
    import numpy as np

    from collections import defaultdict
    from tqdm import tqdm

    ENTITY = 'user'
    PROJECT = 'my_project'
    METRIC_NAME = 'valid/epoch/my_metric'

    api = wandb.Api()

    runs = api.runs(f"{ENTITY}/{PROJECT}")

    for run in tqdm(runs):

        values = run.history()[METRIC_NAME].values
        values = values[~np.isnan(values)]

        run.summary[f"{METRIC_NAME}_max"] = np.max(values)
        run.summary[f"{METRIC_NAME}_min"] = np.min(values)
        run.summary[f"{METRIC_NAME}_std"] = np.std(values)

        run.summary.update()

**Comment by sehoffmann** [comment]
Seems weird to me to have a framework specifically designed for ML experiment tracking and not provide this very essential feature out of the box. Especially since summary stats like min,max,means, quantiles etc. are pretty much the bread and butter for practitioners working with data.

**Comment by LspongebobJH** [comment]
So surprising to me it's been two years and such useful feature still not available.

**Comment by jwh97nn** [comment]
By any chance, is it possible to deal with multiple metrics? Because for some tasks there will be more than one metric, and sorting all metrics based on the main metric will be helpful.
It seems that `define_metric` can only define min/max values for metrics independently, so it may result in the summarized values are not in the same iteration/epoch.

**Comment by RylanSchaeffer** [comment]
How can one modify @SteffenCzolbe 's expression to plot min loss as a function of step?

**Comment by anmolmann** [comment]
Hi everyone, we've recently implemented and released this feature. How to use it -- for any new project, you can now select minimum or maximum run metrics facilitating discovery and analysis using the Runs Table. Please feel free reopen this issue in case of any further queries.

**Comment by wenwenyu** [comment]
> Hi everyone, we've recently implemented and released this feature. How to use it -- for any **_new project_**, you can now select minimum or maximum run metrics facilitating discovery and analysis using the Runs Table. Please feel free reopen this issue in case of any further queries.
> 
> ![2023-08-28 13 15 28](https://user-images.githubusercontent.com/20401632/267412649-4268f357-d7eb-4791-bfc0-35936bd6c90c.gif) [ ![2023-08-28 13 15 28](https://user-images.githubusercontent.com/20401632/267412649-4268f357-d7eb-4791-bfc0-35936bd6c90c.gif) ](https://user-images.githubusercontent.com/20401632/267412649-4268f357-d7eb-4791-bfc0-35936bd6c90c.gif) [ ](https://user-images.githubusercontent.com/20401632/267412649-4268f357-d7eb-4791-bfc0-35936bd6c90c.gif)

Hi, @anmolmann 

Why my dashboard cannot see this option about 'show min'. Is there has any requirements?

**Comment by Lily-Le** [comment]
Can we add a new panel that can show the best results according to a specific chart?  Just for the old runs' sake. Thanks

**Comment by alexaatm** [comment]
Hi, saw that this feature is available in the UI!  :) 
However, only for min and max. Would also be nice to add **STD**  - was what I originally wanted to obtain and how I stumbled on this issue...

Meanwhile, will have to traverse all runs and add it, following the suggestion from @MarkZaretskiy , thank you!

**Comment by ogencoglu** [comment]
<img width="439" alt="Image" src="https://github.com/user-attachments/assets/9678f8ea-a59d-48fd-bb5c-aecfa743228c" />

Is there any way to show the min/max in the plots? Any indication would be fine (circle around the point, horizontal line, vertical line)?

**Comment by Skelp** [comment]
I'm honestly just as surprised as @LspongebobJH to see that I am in fact not blind and this feature is not part of the interface. It is great to see that it was implemented to some extend, but I agree that the most use will come from expressions (akin to what @jacksonthall22 mentioned). I suppose it would be worth trying to implement this and creating a PR for it.
I might give it a shot, but for that it'd be great to get some ideas on this - how do you think it should work syntax-wise, and other than min, max and std should be present?
